### PR TITLE
FIM permutations

### PR DESCRIPTION
**Pull request by Stillerman**
Here is a draft that adds optional FIM permutations while finetuning. Some things I was hoping to get feedback on

- I added the permuting as part of the `ConstantLengthDataset` class. Is that the right place to do this?

- With FIM enabled, batches take ~2x as long to load. Is this acceptable? Here are the stats for loading 5 batches with and without FIM on my Intel MacBook Pro
```
FIM time: 66.68208975399999
Non-FIM time: 32.142377584
```

- I cannot actually test that the training still works on the script because I don't have access to a GPU. The training seems to work in the [Colab I adapted](https://colab.research.google.com/drive/1CADU0YxD3p7pLivgST0toYFLUdGWDP5f) but this should be checked to make sure all is still well

### CODE REVIEW

**Comment by arjunguha** [comment]
I've read this and it seems okay. But, someone should actually run it on a GPU.

**Comment by Stillerman** [comment]
Switched from character-based FIM to token-based. Now we do not have to do expensive encode/decode and it is not slowed down at all.

```
FIM time: 37.271945165999995
Non-FIM time: 37.641369552
```

Will test on GPU soon

**Comment by Stillerman** [comment]
Finetuned for 48 hours on an A100 on the ruby stack and got [this](https://huggingface.co/stillerman/santacoder-ruby).

MultiPL-E pass@1 was 0.1

File: train.py [hunk 1]
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
 
     return total_characters / total_tokens
 
+# this is expensive so we cache it
```
Comment (loubnabnl) [inline train.py]: I suggest moving the `get_fim_token_ids` and `permute` functions to a new file `fim_utils.py` to keep this file short and clear.

File: train.py [hunk 2]
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
...
+
+
+## Adapted from https://github.com/bigcode-project/Megatron-LM/blob/6c4bf908df8fd86b4977f54bf5b8bd4b521003d1/megatron/data/gpt_dataset.py
+def permute(sample, np_rng, tokenizer, fim_rate=0.5, fim_spm_rate=0.5, truncate_or_pad=False):
+    """
+    Take in a sample (np array w/ size (0,chunklength)) and perform a FIM transformation on it. 
+    Maintain the same sample length (if transform creates a few extra tokens, drop them).
+    """
+
+    suffix_tok_id, prefix_tok_id, middle_tok_id, pad_tok_id, = get_fim_token_ids(tokenizer)
```
Comment (loubnabnl) [inline train.py]: we can call this function once in the dataset class and pass values as args to `permute` to avoid calling it for each sample

File: train.py [hunk 3]
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
...
+
+    if np_rng.binomial(1, fim_rate): # sample bernoulli dist
+        try:
+            # A boundary can be =0 (prefix will be empty)
+            # a boundary can be =len(contents) (suffix will be empty)
+            # The two boundaries can be equal (middle will be empty)
```
Comment (loubnabnl) [inline train.py]:
```suggestion
    if np_rng.binomial(1, fim_rate):
        try:
```

File: train.py [hunk 4]
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
...
+            boundaries = list(np_rng.randint(low=0, high=len(sample) + 1, size=2))
+            boundaries.sort()
+        except ValueError as e:
```
Comment (loubnabnl) [inline train.py]: does this ever happen in the training?

File: train.py [hunk 5]
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
...
+            print(e)
+            raise e
+
+        prefix = np.array(sample[:boundaries[0]], dtype=np.int64)
+        middle = np.array(sample[boundaries[0]:boundaries[1]], dtype=np.int64)
+        suffix = np.array(sample[boundaries[1]:], dtype=np.int64)
+
+        # here we truncate each given segment to fit the same length as it was before
+        # A consequence is that we never reach the end of a file?
+        # we should rather truncate at the context-level
```
Comment (loubnabnl) [inline train.py]: we can probably remove these comments if we're not implementing it
```suggestion
```

File: train.py [hunk 6]
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
...
+        if truncate_or_pad:
+            # need to make same length as the input. Take the 3 sentinel tokens into account
+            new_length = suffix.shape[0] + prefix.shape[0] + middle.shape[0] + 3
+            diff = new_length - len(sample)
+            if diff > 0: # too long
+                if suffix.shape[0] <= diff: # if there's no space to truncate the suffix: stop and report it. atm i should have stopped this from happening
```
Comment (loubnabnl) [inline train.py]: same comment as above
```suggestion
                if suffix.shape[0] <= diff:
```

File: train.py [hunk 7]
```diff
@@ -119,7 +194,13 @@ def __iter__(self):
                         break
             tokenized_inputs = self.tokenizer(buffer, truncation=False)["input_ids"]
             all_token_ids = []
+
+            np_rng = np.random.RandomState(seed=555)
```
Comment (loubnabnl) [inline train.py]: let's use the `seed` provided in the arguments and pass it as input to the class

File: train.py [hunk 8]
```diff
@@ -169,8 +252,21 @@ def create_datasets(tokenizer, args):
         seq_length=args.seq_length,
         chars_per_token=chars_per_token,
         content_field=args.data_column,
+        fim_rate=args.fim_rate,
+        fim_spm_rate=args.fim_spm_rate
     )
-    return train_dataset, valid_dataset
+
+    valid_dataset_no_fim = ConstantLengthDataset(
```
Comment (loubnabnl) [inline train.py]: Did you want to add an option to compute validation loss in non FIM mode? Currently this dataset is not used in the script

File: train.py [hunk 9]
```diff
@@ -219,7 +315,14 @@ def run_training(args, train_data, val_data):
 
 def main(args):
     tokenizer = AutoTokenizer.from_pretrained(args.model_path, use_auth_token=True)
-    train_dataset, eval_dataset = create_datasets(tokenizer, args)
+    train_dataset, eval_dataset, eval_no_fim = create_datasets(tokenizer, args)
```
Comment (loubnabnl) [inline train.py]: `eval_no_fim` is not used (see comment above). We can add an option if the user wants to evaluate on both datasets when `fim_rate` is not zero. To do that, you can pass a `dict` as `eval_dataset` to the `Trainer` ([see](eval_no_fim))

File: train.py [hunk 10]
```diff
@@ -219,7 +315,14 @@ def run_training(args, train_data, val_data):
...
+
+    fim_time = timeit.timeit(lambda: next(iter(eval_dataset)), number=5)
+    non_fim_time = timeit.timeit(lambda: next(iter(eval_no_fim)), number=5)
+
+    print(f"FIM time: {fim_time}")
+    print(f"Non-FIM time: {non_fim_time}")
+
```
Comment (loubnabnl) [inline train.py]: You didn't observe a slow down in CPU nor GPU, right? If so we can remove this
```suggestion
```

File: train.py [hunk 11]
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
...
+def get_fim_token_ids(tokenizer):
+    _, FIM_PREFIX, FIM_MIDDLE, FIM_SUFFIX, FIM_PAD = tokenizer.special_tokens_map['additional_special_tokens']
+    suffix_tok_id, prefix_tok_id, middle_tok_id, pad_tok_id = (tokenizer.vocab[tok] for tok in [FIM_SUFFIX, FIM_PREFIX, FIM_MIDDLE, FIM_PAD])
+    return suffix_tok_id, prefix_tok_id, middle_tok_id, pad_tok_id
+
+
+## Adapted from https://github.com/bigcode-project/Megatron-LM/blob/6c4bf908df8fd86b4977f54bf5b8bd4b521003d1/megatron/data/gpt_dataset.py
+def permute(sample, np_rng, tokenizer, fim_rate=0.5, fim_spm_rate=0.5, truncate_or_pad=False):
+    """
+    Take in a sample (np array w/ size (0,chunklength)) and perform a FIM transformation on it.
```
Comment (loubnabnl) [inline train.py]:
```suggestion
    Take in a sample (list of tokens) and perform a FIM transformation on it with a probability of fim_rate, using two FIM modes: PSM and SPM (with a probability of fim_spm_rate). 
```

**Review by loubnabnl — COMMENTED** [review]
Thanks a lot for this contribution! The code looks good, I just left some small comments mainly about the format. 
Let's move the FIM functions to a separate file to keep the main file concise and easy. Can you also run `isort` and `black` on the file?

File: train.py (hunk 1 above)
Comment (Stillerman) [reply train.py]: done

File: train.py (hunk 2 above)
Comment (Stillerman) [reply train.py]: done

File: train.py (hunk 3 above)
Comment (Stillerman) [reply train.py]: 👍

File: train.py (hunk 4 above)
Comment (Stillerman) [reply train.py]: Not that I saw, and I can't imagine how it would. Removed.

File: train.py (hunk 5 above)
Comment (Stillerman) [reply train.py]: 👍

File: train.py (hunk 6 above)
Comment (Stillerman) [reply train.py]: 👍

File: train.py (hunk 7 above)
Comment (Stillerman) [reply train.py]: 👍

File: train.py (hunk 9 above)
Comment (Stillerman) [reply train.py]: That was left over from testing. I don't see why we would want non-FIM eval so I have removed for now. Is this something we would want?

File: train.py (hunk 10 above)
Comment (Stillerman) [reply train.py]: Removed

File: train.py (hunk 11 above)
Comment (Stillerman) [reply train.py]: 👍

**Comment by Stillerman** [comment]
I addressed all of the comments in that last commit. The only change really of note is that I removed the `eval_no_fim` instead of making that an additional eval set, as that was left over from testing. Is there some reason why we would want to add it as an additional eval set?

Also, reran a toy finetune for an hour and everything seems to be working same as it did before.

**Comment by Stillerman** [comment]
Finetuned Julia for 2m tokens both with and without FIM and tested both with MultiPL-E.

[With FIM](https://huggingface.co/stillerman/santacoder-julia-fim) pass@1 = 0.1

[Without FIM](https://huggingface.co/stillerman/santacoder-julia-no-fim) pass@1 = 0.09

**Comment by loubnabnl** [comment]
Looks great! Thanks for working on this and for your patience!
